File: CucumberJVM/Chapter13/13.1-TheLocalScenarioContextPattern/Before/src/main/java/com/wimp/localscenariocontext/app/infrastructure/DataContext.java
package com.wimp.localscenariocontext.app.infrastructure;

import com.wimp.localscenariocontext.app.models.Notification;
import com.wimp.localscenariocontext.app.models.Order;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * A simple simulation of a database context, using in-memory collections. In a real application,
 * this would likely be replaced with an actual database context or a repository pattern.
 */
public class DataContext {
    public static final DataContext INSTANCE = new DataContext();

    private final Map<Integer, Order> orders = new HashMap<>();
    private final List<Notification> notifications = new ArrayList<>();

    public void reset() {
        orders.clear();
        notifications.clear();
    }

    public List<Notification> getNotificationsByCustomerName(String customerName) {
        return notifications.stream()
                .filter(n -> n.getCustomerName().equals(customerName))
                .collect(Collectors.toList());
    }

    public void saveNotification(Notification notification) {
        notifications.add(notification);
    }

    public Order getOrderByOrderNr(int orderNr) {
        return orders.values().stream()
                .filter(o -> o.getOrderNo() == orderNr)
                .findFirst()
                .orElse(null);
    }

    public void saveOrder(Order order) {
        orders.put(order.getOrderNo(), order);
    }
}

File: Reqnroll/Chapter13/13.2-TheShareableScenarioContextPattern/After/WIMP.ShareableScenarioContextSample.App/Infrastructure/DataContext.cs
using WIMP.ShareableScenarioContextSample.App.Models;

namespace WIMP.ShareableScenarioContextSample.App.Infrastructure;

/// <summary>
/// A simple simulation of a database context, using in-memory collections. In a real application,
/// this would likely be replaced with an actual database context (e.g., Entity Framework DbContext) or a repository pattern.
/// </summary>
public class DataContext
{
    public static readonly DataContext Instance = new();

    private readonly Dictionary<int, Order> orders = new();
    private readonly List<Notification> notifications = new();

    public void Reset()
    {
        orders.Clear();
        notifications.Clear();
    }

    public IEnumerable<Notification> GetNotificationsByCustomerName(string customerName) =>
        notifications.Where(n => n.CustomerName == customerName);

    public void SaveNotification(Notification notification)
    {
        notifications.Add(notification);
    }

    public Order? GetOrderByOrderNr(int orderNr) =>
        orders.Values.FirstOrDefault(o => o.OrderNo == orderNr);

    public void SaveOrder(Order order)
    {
        orders.Add(order.OrderNo, order);
    }
}
